addons/l10n_it_edi_sdicoop/models/res_config_settings.py

```diff
@@ -39,3 +62,40 @@ def button_create_proxy_user(self):
             return
 
         self.env['account_edi_proxy_client.user']._register_proxy_user(self.company_id, fattura_pa, edi_identification)
+
+    def _set_l10n_it_edi_sdicoop_register_demo_mode(self):
+
+        fattura_pa = self.env.ref('l10n_it_edi.edi_fatturaPA')
+        for config in self:
+
+            proxy_user = self.env['account_edi_proxy_client.user'].search([
+                ('company_id', '=', config.company_id.id),
+                ('edi_format_id', '=', fattura_pa.id)
+            ], limit=1)
+
+            real_proxy_users = self.env['account_edi_proxy_client.user'].sudo().search([
+                ('id_client', 'not like', 'demo'),
+            ])
+
+            # Update the config as per the selected radio button
+            previous_demo_state = proxy_user._get_demo_state()
+            self.env['ir.config_parameter'].set_param('account_edi_proxy_client.demo', config.l10n_it_edi_sdicoop_demo_mode)
+
+            # If the user is trying to change from a state in which they have a registered official or testing proxy client
+            # to another state, we should stop them
+            if real_proxy_users and previous_demo_state != config.l10n_it_edi_sdicoop_demo_mode:
+                raise UserError(_("The company has already registered with the service as 'Test' or 'Official', it cannot change."))
+
+
+            if config.l10n_it_edi_sdicoop_register:
+                # There should only be one user at a time, if there are no users, register one
+                if not proxy_user:
+                    self._create_proxy_user(config.company_id)
+                    return
+
+                # If there is a demo user, and we are transitioning from demo to test or production, we should
+                # delete all demo users and then create the new user.
+                elif proxy_user.id_client[:4] == 'demo' and config.l10n_it_edi_sdicoop_demo_mode != 'demo':
+                    self.env['account_edi_proxy_client.user'].search([('id_client', '=like', 'demo%')]).sudo().unlink()
+                    self._create_proxy_user(config.company_id)
+                    return
```

Comment:
@dbkosky Could we remove this last return and put a newline?  (it would avoid ci/style errors in master)  Otherwise, it works fine.